functions/go/source-gcloud-config/gcloudconfig/generator.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package gcloudconfig
+
+import (
+	"github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/source-gcloud-generator/exec"
+	"sigs.k8s.io/kustomize/kyaml/kio/filters"
+	"sigs.k8s.io/kustomize/kyaml/resid"
+
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const GcloudMetaName = "gcloud-config.kpt.dev"
+
+var LocalConfigAnnos = map[string]string{filters.LocalConfigAnnotation: "true"}
+
+// GcloudConfigGenerator is the generator function to generate a GcloudConfig resource.
+type GcloudConfigGenerator struct{}
+
+// Generate executes `gcloud` commands to create a GcloudConfig RNode.
+func (g *GcloudConfigGenerator) Generate(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	gcloudData, err := exec.GetGcloudContextFn()
+	if err != nil {
+		return nil, err
+	}
+	gcloudNode, err := NewGcloudConfigNode(gcloudData)
+	if err != nil {
+		return nil, err
+	}
+	var newNodes []*yaml.RNode
+	for _, curNode := range nodes {
+		gcloudGvk := resid.GvkFromNode(gcloudNode)
+		curNodeGvk := resid.GvkFromNode(curNode)
+		if curNode.GetName() == gcloudNode.GetName() && curNodeGvk.Equals(gcloudGvk) {
+			continue
+		}
+		newNodes = append(newNodes, curNode)
+	}
+	newNodes = append(newNodes, gcloudNode)
```

Comment:
Since this is pure gcloud context and does not contain Kptfile or kpt package information, the `gcloud-config` resources will always be the same, and the the gcloudNode is stored on the top-level package. 

Do you mind explaining more about "they will simply be dropped and only one will be added at the top". I was assuming "drop" means the sub-packge gcloud-config.yaml files are removed from disk. But my manual testing shows sub-package gcloud-config files are not passed into the hydration pipeline. I doubt if that's because all gcloud-config shares the same GVKNN so duplicate nodes are filtered out. I haven't dug into the code to prove that. But by running `kpt fn source` only, it seems only the top-packge gcloud-config is listed.

Reply:
>  I was assuming "drop" means the sub-packge gcloud-config.yaml files are removed from disk. But my manual testing shows sub-package gcloud-config files are not passed into the hydration pipeline.

It is because `kpt fn render` doesn't provide `meta resources` as input to functions today. This PR https://github.com/GoogleContainerTools/kpt/pull/2781 is going to change that.

Reply:
This gcloud-config is `ConfigMap` kind, when not used as functionConfig, running `kpt fn source` only gives me the top level gcloud-config and ignore the gcloud-config from sub-package.

Reply:
yeah, the "weird" behavior is because the gcloud is passed in as meta resource

Since all gcloud-config.yaml have the same gcloud context, I assume running this gcloud config function shall update both top package (create if not exists) and  sub pacakges gcloud-config.yaml file. I updated the logic to reflect that. Please see the last commit